pkg/virt-controller/watch/vm_initializer.go

```diff
@@ -0,0 +1,371 @@
+			continue
+		}
+		logger.Infof("VirtualMachinePreset %s matches labels of VM %s", preset.GetName(), vm.GetName())
+		matchingPresets = append(matchingPresets, preset)
+	}
+	return matchingPresets, nil
+}
+
+func checkPresetMergeConflicts(presetSpec *kubev1.DomainSpec, vmSpec *kubev1.DomainSpec) error {
+	errors := []error{}
+	if len(presetSpec.Resources.Requests) > 0 {
+		for key, presetReq := range presetSpec.Resources.Requests {
+			if vmReq, ok := vmSpec.Resources.Requests[key]; ok {
+				if presetReq != vmReq {
+					errors = append(errors, fmt.Errorf("spec.resources.requests[%s]: %v != %v", key, presetReq, vmReq))
+				}
+			}
+		}
+	}
+	if presetSpec.CPU != nil && vmSpec.CPU != nil {
+		if !reflect.DeepEqual(presetSpec.CPU, vmSpec.CPU) {
+			errors = append(errors, fmt.Errorf("spec.cpu: %v != %v", presetSpec.CPU, vmSpec.CPU))
+		}
+	}
+	if presetSpec.Firmware != nil && vmSpec.Firmware != nil {
+		if !reflect.DeepEqual(presetSpec.Firmware, vmSpec.Firmware) {
+			errors = append(errors, fmt.Errorf("spec.firmware: %v != %v", presetSpec.Firmware, vmSpec.Firmware))
+		}
+	}
+	if presetSpec.Clock != nil && vmSpec.Clock != nil {
+		if !reflect.DeepEqual(presetSpec.Clock.ClockOffset, vmSpec.Clock.ClockOffset) {
+			errors = append(errors, fmt.Errorf("spec.clock.clockoffset: %v != %v", presetSpec.Clock.ClockOffset, vmSpec.Clock.ClockOffset))
+		}
+		if presetSpec.Clock.Timer != nil && vmSpec.Clock.Timer != nil {
+			if !reflect.DeepEqual(presetSpec.Clock.Timer, vmSpec.Clock.Timer) {
+				errors = append(errors, fmt.Errorf("spec.clock.timer: %v != %v", presetSpec.Clock.Timer, vmSpec.Clock.Timer))
+			}
+		}
+	}
+	if presetSpec.Features != nil && vmSpec.Features != nil {
+		if !reflect.DeepEqual(presetSpec.Features, vmSpec.Features) {
+			errors = append(errors, fmt.Errorf("spec.features: %v != %v", presetSpec.Features, vmSpec.Features))
+		}
+	}
+	if presetSpec.Devices.Watchdog != nil && vmSpec.Devices.Watchdog != nil {
+		if !reflect.DeepEqual(presetSpec.Devices.Watchdog, vmSpec.Devices.Watchdog) {
+			errors = append(errors, fmt.Errorf("spec.devices.watchdog: %v != %v", presetSpec.Devices.Watchdog, vmSpec.Devices.Watchdog))
+		}
+	}
+	nameMap := make(map[string]kubev1.Disk)
+	volumeNameMap := make(map[string]kubev1.Disk)
+
+	for _, vmDev := range vmSpec.Devices.Disks {
+		nameMap[vmDev.Name] = vmDev
+		volumeNameMap[vmDev.VolumeName] = vmDev
+	}
+	for _, presetDev := range presetSpec.Devices.Disks {
+		if vmDev, conflict := nameMap[presetDev.Name]; conflict {
+			if !reflect.DeepEqual(presetDev, vmDev) {
+				errors = append(errors, fmt.Errorf("spec.devices.disk[%s]: conflicting disk with same name", presetDev.Name))
+			}
+		}
+		if vmDev, conflict := volumeNameMap[presetDev.VolumeName]; conflict {
+			if !reflect.DeepEqual(presetDev, vmDev) {
+				errors = append(errors, fmt.Errorf("spec.devices.disk[%s]: conflicting disk with same volume name", presetDev.Name))
+			}
+		}
+	}
+	if len(errors) > 0 {
+		return utilerrors.NewAggregate(errors)
```

Comment:
each of these conflict errors need to be posted as an event on the VM object. Look at the "recorder" object in the vm controller to see how this is done.

Reply:
You are correct! I believe this was supposed to be addressed in a follow-on PR.